File: app/src/main/java/xyz/untan/gyazauto/ShareActivity.java
package xyz.untan.gyazauto;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

public class ShareActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent = getIntent();
        String action = intent.getAction();
        String type = intent.getType();
        Uri imageUri = intent.getParcelableExtra(Intent.EXTRA_STREAM);

        if (!(action != null && action.equals(Intent.ACTION_SEND)
                && type != null && type.startsWith("image/")
                && imageUri != null)) {
            return;
        }

        // start UploadService
        intent = new Intent(this, UploadService.class);
        intent.setType(type);
        intent.putExtra(Intent.EXTRA_STREAM, imageUri);

        startService(intent);

        finish(); // no need to show activity
    }
}

File: app/src/main/java/xyz/untan/gyazauto/ShareActivity.kt
package xyz.untan.gyazauto

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity

class ShareActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val intent: Intent? = intent
        if (intent == null) {
            finish()
            return
        }

        val imageUri: Uri? = intent.getParcelableExtra<Uri>(Intent.EXTRA_STREAM)

        if (!(intent.action == Intent.ACTION_SEND && intent.type?.startsWith("image/") ?: false
                && imageUri != null)) {
            return
        }

        // start UploadService
        val newIntent = Intent(this, UploadService::class.java).apply {
            type = intent.type
            putExtra(Intent.EXTRA_STREAM, imageUri)
        }

        startService(newIntent)

        finish() // no need to show activity
    }
}
